Source/JavaScriptCore/runtime/JSGenericTypedArrayViewPrototype.cpp

```diff
@@ -0,0 +1,106 @@
+ * THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL APPLE INC. OR ITS CONTRIBUTORS
+ * BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "config.h"
+#include "JSGenericTypedArrayViewPrototype.h"
+
+#include "NumberPrototype.h"
+#include <wtf/text/Base64.h>
+
+namespace JSC {
+
+JSC_DEFINE_HOST_FUNCTION(uint8ArrayPrototypeToBase64, (JSGlobalObject* globalObject, CallFrame* callFrame))
+{
+    VM& vm = globalObject->vm();
+    auto scope = DECLARE_THROW_SCOPE(vm);
+
+    JSValue thisValue = callFrame->thisValue();
+    if (!thisValue.isObject())
+        return throwVMTypeError(globalObject, scope, "Receiver should be a typed array view but was not an object"_s);
+
+    JSObject* thisObject = jsCast<JSObject*>(thisValue);
+    if (thisObject->type() != Uint8ArrayType)
+        return throwVMTypeError(globalObject, scope, "Uint8Array.prototype.toBase64 requires that |this| be a Uint8Array"_s);
+
+    JSUint8Array* uint8Array = jsCast<JSUint8Array*>(thisObject);
+    validateTypedArray(globalObject, uint8Array);
+
+    Base64EncodeMode mode = Base64EncodeMode::Default;
+
+    JSValue optionsValue = callFrame->argument(0);
+    if (!optionsValue.isUndefined()) {
+        if (!optionsValue.isObject())
+            return throwVMTypeError(globalObject, scope, "Uint8Array.prototype.toBase64 requires that options be an object"_s);
+
+        JSObject* optionsObject = jsCast<JSObject*>(optionsValue);
+
+        JSValue alphabetValue = optionsObject->get(globalObject, Identifier::fromString(vm, "alphabet"_s));
+        RETURN_IF_EXCEPTION(scope, { });
+        if (!alphabetValue.isUndefined()) {
+            if (!alphabetValue.isString())
+                return throwVMTypeError(globalObject, scope, "Uint8Array.prototype.toBase64 requires that alphabet be \"base64\" or \"base64url\""_s);
+
+            String alphabet = alphabetValue.toWTFString(globalObject);
+            if (alphabet == "base64url"_s)
+                mode = Base64EncodeMode::URL;
+            else if (alphabet != "base64"_s)
+                return throwVMTypeError(globalObject, scope, "Uint8Array.prototype.toBase64 requires that alphabet be \"base64\" or \"base64url\""_s);
+        }
+    }
+
+    const uint8_t* data = uint8Array->typedVector();
+    size_t offset = uint8Array->byteOffset();
+    size_t length = uint8Array->length();
+    RELEASE_AND_RETURN(scope, JSValue::encode(jsString(vm, base64EncodeToString({ data + offset, length }, mode))));
```

Comment:
This is not correct. `typedVector` is already offseted vector pointer. `byteOffset` is indicating the offset in the original buffer. Not in the typedVector. Can you add a test for that?